### ISSUE TITLE
Document guidelines on how to annotate methods with availability annotations

### ISSUE THREAD

**Issue by jverkoey** [material-components/material-components-ios#7070]
We are increasingly starting to make use of availability annotations for our APIs. In order to reduce churn and inconsistent feedback on PRs, we should document a team practice that we can point to.

A recommended practice:

> Any private method that solely consists of logic nested within an `if (@available(` statement and is not a subclass override should be annotated with NS_AVAILABLE_IOS instead. This allows us to remove a layer of logical indentation and more clearly communicates that the method itself is only available for a given platform version.

---
#### Internal data
- Associated internal bug: [b/129901187](http://b/129901187)

**Comment by romoore** [comment]
I would suggest that installing runtime guards at the caller (rather than the callee) makes sense if there is an alternative branch.  If the guarded code is an all-or-nothing situation, then it's better to put the runtime guard inside the callee to prevent mistakes.

I believe this is how we've implemented `mdc_safeAreaInsets` and some other "backported" APIs that would have no effect (but are safe to call) on earlier versions.

## Example

**Caller has an alternative to the guarded code:**

```objc
- (void)caller {
  if (@available(iOS 9.0, *)) {
    [self doiOS9Thing];
  } else {
    [self doPreiOS9Thing];
  }
}

- (void)doiOS9Thing NS_AVAILABLE_IOS(9_0) {
  [self iOS9OnlyMethod];
}
```

**Caller has no alternative to the guarded code:**

```objc
- (void)caller {
  [self doiOS9ThingOrNothing];
}

- (void)doiOS9ThingOrNothing {
  if (@available(iOS 9.0, *)) {
    [self iOS9OnlyMethod];
  }
}
```

**Comment by jverkoey** [comment]
My concern with the proposed `doiOS9ThingOrNothing` proposal is that we do not document in our APIs that methods do nothing on specific platforms, so there may be expectations that a method does, in fact, implement some level of fall-back behavior.

In cases like this, I believe that it a clear annotation on the API results in clearer enforcement of the expected behavior of the method.
